### PR TITLE
Bug 1669566 add cookie store id to browser.downloads

### LINKED ISSUE #7747: [webEXT] Add cookieStoreId option to browser.downloads APIs
[Add cookieStoreId to browser.downloads](https://bugzilla.mozilla.org/show_bug.cgi?id=1669566)

## Acceptance Criteria

- [ ] The listed features are documented sufficiently on MDN
- [ ] BCD is updated
- [ ] Interactive example and data repos are updated if appropriate
- [ ] The content has been reviewed as needed

### For folks helping with Firefox related documentation

- [ ] Set bugs to `dev-doc-complete`
- [ ] Add entry to Firefox release notes if feature is enabled in release
or
- [ ] Add entry to Firefox experimental features page if feature is not yet enabled in release

## Features to document

- Add cookieStoreId option to browser.downloads APIs

## Related Gecko bugs

- [browser.downloads.download method only passes default container cookies](https://bugzilla.mozilla.org/show_bug.cgi?id=1669566).

### PR DESCRIPTION
**Pull request by rebloor**
Changes adding cookieStoreId to browser.download.download, downloads.DownloadQuery (browser.download.search), and downloads.DownloadItem (browser.download.erase). This provides documentation for  [Bug 1669566](https://bugzilla.mozilla.org/show_bug.cgi?id=1669566).

Corresponding changes to browser-compat-data Are in [Bug-1669566-Add-cookieStoreId-to-browser.downloads #12069](https://github.com/mdn/browser-compat-data/pull/12069) .

### CODE REVIEW

File: files/en-us/mozilla/add-ons/webextensions/api/downloads/download/index.html [hunk 1]
```diff
@@ -49,6 +49,9 @@ <h3 id="Parameters">Parameters</h3>
   <dd>A <code>string</code> representing the post body of the request.</dd>
   <dt><code>conflictAction</code>{{optional_inline}}</dt>
   <dd>A string representing the action you want taken if there is a filename conflict, as defined in the {{WebExtAPIRef('downloads.FilenameConflictAction')}} type (defaults to "uniquify" when it is not specified).</dd>
+  <dt><code>cookieStoreId</code>{{optional_inline}}</dt>
+  <dd>The cookie store ID of the contextual identity the download is associated with. If omitted, the default cookie store is used. Use requires the "cookies" <a href="/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#api_permissions">API permission</a>.
```
Comment (hamishwillee) [inline files/en-us/mozilla/add-ons/webextensions/api/downloads/download/index.html:53]: Is "contextual identity" enough for someone who is somewhat familiar with web extensions to know what you mean?

**Comment by hamishwillee** [comment]
The original bug referred to the cookieStoreId applying to `erase` as well. I presume that ended up being documented in the DownloadQuery?

**Comment by rebloor** [comment]
> The original bug referred to the cookieStoreId applying to `erase` as well. I presume that ended up being documented in the DownloadQuery?

Yes,  because `browser.download.erase` and `browser.download.search` both use `downloads.DownloadQuery` the "addition" was achieved by adding `cookieStoreId` to the type.  a similar reason to why `cookieStoreId` was added to `downloads.DownloadItem`.

File: files/en-us/mozilla/add-ons/webextensions/api/downloads/download/index.html (hunk 1 above)
Comment (rebloor) [reply files/en-us/mozilla/add-ons/webextensions/api/downloads/download/index.html]: I've added a cross-reference link to the article on how to work with contextual identities, which should help anyone who's unfamiliar with this. I also noticed that there isn't any reference to contextual identities in the details about the cookies API and have opened a [new issue about this](https://github.com/mdn/content/issues/8267).

I've not been able to find any information specifically relating to how cookies are used with downloads. My assumption is that the behavior is identical to that for web pages e.g. when the download occurs the website may create or read cookies. Similarly, the extension may create or read cookies to store its own information about the download. @Rob--W you commented on the original bug, is there anything we need to add about cookie use/behavior here?

I have to admit I find
> "If the specified url uses the HTTP or HTTPS protocol, then the request will include all cookies currently set for its hostname."
a little ambiguous.  I assume this means that any cookies set for the hostname could be updated by the download host as part of the download. @Rob--W could you help clarify?

**Review by hamishwillee — CHANGES_REQUESTED** [review]
This is good, but I think this query is still open: https://github.com/mdn/content/pull/8106/files#discussion_r695064608

Specifically, up the top it says

> If the specified url uses the HTTP or HTTPS protocol, then the request will include all cookies currently set for its hostname.

And in the `cookieStoreId` it says

> If omitted, the default cookie store is used.". 

My assumption is that the top point is the "default", but if you use the `cookieStoreId` then you get a different set. However it might be that you get the `cookieStoreId` as well. I think both places should be explicit - define the default set and up the top note that the specified URL might be overridden.

But really up to Rob to confirm

File: files/en-us/mozilla/firefox/releases/92/index.html [hunk 2]
```diff
@@ -73,6 +73,10 @@ <h4 id="removals_webdriver">Removals</h4>
 
 <h2 id="Changes_for_add-on_developers">Changes for add-on developers</h2>
 
+<ul>
+  <li>Support for <code>cookieStoreId</code> is added to {{WebExtAPIRef('downloads.download')}}, {{WebExtAPIRef('downloads.DownloadQuery')}}, and {{WebExtAPIRef('downloads.DownloadItem')}}. The addition to the two types provides support in {{WebExtAPIRef('downloads.search')}} and {{WebExtAPIRef('downloads.erase')}}. Web extensions using <a href="/en-US/docs/Mozilla/Add-ons/WebExtensions/Work_with_contextual_identities">contextual identities</a> can now use the appropriate cookie store when performing downloads. ({{bug(1669566)}})</li>
```
Comment (hamishwillee) [inline files/en-us/mozilla/firefox/releases/92/index.html:77]: This is strictly accurate. FWIW I did find the text "The addition to the two types provides support in" a little hard to parse, because it wasn't immediately obvious (to me) that the preceding sentence was one method and two types. Probably not worth "fixing".

Comment (rebloor) [reply files/en-us/mozilla/firefox/releases/92/index.html:77]: Methods all start with a lowercase letter, types with an uppercase letter, properties are either block capitals or start with a lowercase letter, and events prefixed with "on".  If we need changes for the other outstanding point, I'll update.

File: files/en-us/mozilla/add-ons/webextensions/api/downloads/download/index.html (hunk 1 above)
Comment (Rob--W) [reply files/en-us/mozilla/add-ons/webextensions/api/downloads/download/index.html]: > I've not been able to find any information specifically relating to how cookies are used with downloads. My assumption is that the behavior is identical to that for web pages e.g. when the download occurs the website may create or read cookies. 

It is indeed:
- If the `incognito` option is used, private browsing cookies are used.
- If the `cookieStoreId` option is used, the cookie store described by this parameter is used.
- If neither is given, the default cookies are used (i.e. from the normal browsing session).


> Similarly, the extension may create or read cookies to store its own information about the download. @Rob--W you commented on the original bug, is there anything we need to add about cookie use/behavior here?

Although the extension offers a `headers` option, the option does not allow custom cookies to be set. Only existing cookies are passed with the request (and any cookies received from the server will be saved if any).


> I have to admit I find
> 
> > "If the specified url uses the HTTP or HTTPS protocol, then the request will include all cookies currently set for its hostname."
> 
> a little ambiguous. I assume this means that any cookies set for the hostname could be updated by the download host as part of the download. @Rob--W could you help clarify?

The line here is a technically not incorrect description of how cookies are being sent with requests, but it's very confusing. Cookies are indeed derived from the host name (but also other components such as scheme (secure flag), path, and so on).

Given the confusion here, I think that it's worthwhile to clarify this (e.g. by being less vague and including what I described above if needed).

File: files/en-us/mozilla/firefox/releases/92/index.html (hunk 2 above)
Comment (Rob--W) [reply files/en-us/mozilla/firefox/releases/92/index.html:77]: Use `WebExtensions`, not `Web extensions`.

The emphasis on "extensions using contextual identities" is somewhat unfortunate. It seems to suggest that the feature is only available to extensions that "use" contextual identities (there is a `contextualIdentities` API and permission). "use" could suggest two things: management of identities, or operating a container-aware extension. The current wording and link skews more towards the former, while the feature is intended to allow extensions to be more container-aware.

Could you rephrase it, so that it sounds more like `Extensions can now associate downloads with specific cookie stores, such as container tabs`?

**Review by Rob--W — CHANGES_REQUESTED** [review]
`cookieStoreId` is not only used with contextual identities ("container tabs" is the more common term), but it can also be used to select the private browsing mode or the normal browsing context. Limiting the documentation to just the "contextual identities" feature is not accurate.

I think that the best way to resolve this is to add a new article that describes the `cookieStoreId` concept (https://github.com/mdn/content/issues/8267) and link to that from here.

File: files/en-us/mozilla/firefox/releases/92/index.html (hunk 2 above)
Comment (hamishwillee) [reply files/en-us/mozilla/firefox/releases/92/index.html:77]: And that perhaps confirms https://github.com/mdn/content/pull/8106#pullrequestreview-742151844 - the default is the cookies from the host, but you can use the ones associated with specific cookie stores. Near the top it should say that.

**Review by Rob--W — APPROVED** [review]
The current version of the patch doesn't address the top-level comment that I had at https://github.com/mdn/content/pull/8106#pullrequestreview-743045401, i.e. that `cookieStoreId` is not just limited to containers. In the absence of better documentation, I'm fine with merging this PR as is, but when an article for `cookieStoreId` is added (in #8267), then the `cookieStoreId` property descriptions introduced by this PR should be updated to link to that new article instead.

**Comment by rebloor** [comment]
@hamishwillee Do the changes resolve your change request?

**Review by hamishwillee — APPROVED** [review]
@Rob--W Yes, the bit I asked to be changed has been changed and makes it clear that the  cookies depend on more than just the protocol. 

> If the URL uses the HTTP or HTTPS protocol, the request includes all the relevant cookies, that is, those cookies set for the URL's hostname, secure flag, path, and so on. The default cookies, the cookies from the normal browsing session, are used unless:

The only reason I slightly question it is that it defines "The default cookies" as "the cookies from the normal browsing session" - and I immediately wondered how that differed from " those cookies set for the URL's hostname, secure flag, path, and so on."

I suspect someone more expert than me will get the distinction so no change required.

**Comment by Rumyra** [comment]
Thanks everyone! Going to merge 👍
